tools/stubtest.py

```diff
@@ -0,0 +1,84 @@
+import ast
+import os
+import pathlib
+import subprocess
+import sys
+import tempfile
+
+root = pathlib.Path(__file__).parent.parent
+
+lib = root / "lib"
+mpl = lib / "matplotlib"
+
+
+class Visitor(ast.NodeVisitor):
+    def __init__(self, filepath, output):
+        self.filepath = filepath
+        self.context = list(filepath.with_suffix("").relative_to(lib).parts)
+        self.output = output
+
+    def visit_FunctionDef(self, node):
+        if any("delete_parameter" in ast.unparse(line) for line in node.decorator_list):
+            parents = []
+            if hasattr(node, "parent"):
+                parent = node.parent
+                while hasattr(parent, "parent") and not isinstance(parent, ast.Module):
+                    parents.insert(0, parent.name)
+                    parent = parent.parent
+            self.output.write(f"{'.'.join(self.context + parents)}.{node.name}\n")
+
+    def visit_ClassDef(self, node):
+        for dec in node.decorator_list:
+            if "define_aliases" in ast.unparse(dec):
+                parents = []
+                if hasattr(node, "parent"):
+                    parent = node.parent
+                    while hasattr(parent, "parent") and not isinstance(
+                        parent, ast.Module
+                    ):
+                        parents.insert(0, parent.name)
+                        parent = parent.parent
+                aliases = ast.literal_eval(dec.args[0])
+                # Written as a regex rather than two lines to avoid unused entries
+                # for setters on items with only a getter
+                for substitutions in aliases.values():
+                    parts = self.context + parents + [node.name]
+                    self.output.write(
+                        "\n".join(
+                            f"{'.'.join(parts)}.[gs]et_{a}\n" for a in substitutions
+                        )
+                    )
+        for child in ast.iter_child_nodes(node):
+            self.visit(child)
+
+
+with tempfile.NamedTemporaryFile("wt", delete=False) as f:
```

Comment:
Ensure this cleans up like the context manager:
```suggestion
# Use delete_on_close when Python 3.12 is minimum.
f = tempfile.NamedTemporaryFile("wt", delete=False)
try:
```

Reply:
Alternatively you can just use a TemporaryDirectory instead and create a file with whatever name you want in that directory, to get both proper autodeletion without problems with close()ing the file.  We use that pattern quite a few times in the library itself.

Reply:
Done